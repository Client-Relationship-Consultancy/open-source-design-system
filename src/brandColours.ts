export const colourPalette = {
  examplePalette: {
    primary: {
      main: {
        hex: "#A42824",
        RGB: "164,40,36",
        on: "#FFFFFF",
      },
      light: {
        hex: "#DA342F",
        RGB: "218,52,47",
        on: "#FFFFFF",
      },
      dark: {
        hex: "#661614",
        RGB: "102,22,20",
        on: "#FFFFFF",
      },
    },
    secondary: {
      main: {
        hex: "#0B466B",
        RGB: "11,70,107",
        on: "#FFFFFF",
      },
      light: {
        hex: "#599CAF",
        RGB: "89,156,175",
        on: "#FFFFFF",
      },
      dark: {
        hex: "#05263B",
        RGB: "5,38,59",
        on: "#FFFFFF",
      },
    },
    complimentary: {
      main: {
        hex: "#FFEE52",
        RGB: "255,238,82",
        on: "#272727",
      },
      light: {
        hex: "#FFF59D",
        RGB: "255,245,157",
        on: "#272727",
      },
      dark: {
        hex: "#CEB900",
        RGB: "206,185,0",
        on: "#272727",
      },
    },
    action: {
      main: {
        hex: "#743C8F",
        RGB: "116,60,143",
        on: "#FFFFFF",
      },
      light: {
        hex: "#AD65CF",
        RGB: "173,101,207",
        on: "#FFFFFF",
      },
      dark: {
        hex: "#4E2662",
        RGB: "79,38,98",
        on: "#FFFFFF",
      },
    },
    error: {
      main: {
        hex: "#FF7320",
        RGB: "255,115,32",
        on: "#FFFFFF",
      },
      dark: {
        hex: "#C54901",
        RGB: "197,73,1",
        on: "#FFFFFF",
      },
    },
    background: {
      hex: "#F6F6F6",
      RGB: "246,246,246",
    },
    surface: {
      hex: "#FFFFFF",
      RGB: "255,255,255",
      on: "#272727",
    },
    white: {
      hex: "#FFFFFF",
      RGB: "255,255,255",
      on: "#272727",
    },
    black: {
      main: { hex: "#272727", RGB: "39,39,39", on: "#FFFFFF" },
      tint80: { hex: "#696969", RGB: "105,105,105", on: "#FFFFFF" },
      tint60: { hex: "#A0A0A0", RGB: "160,160,160", on: "#FFFFFF" },
      tint40: { hex: "#C7C7C7", RGB: "199,199,199", on: "#272727" },
      tint20: { hex: "#ECECEC", RGB: "236,236,236", on: "#272727" },
    },
  },
};

interface IColourCodes {
  hex: string;
  RGB: string;
  on?: string;
}
interface IColourVariations {
  main: IColourCodes;
  light?: IColourCodes;
  dark?: IColourCodes;
}

interface IBlackVariations {
  main: IColourCodes;
  tint80: IColourCodes;
  tint60: IColourCodes;
  tint40: IColourCodes;
  tint20: IColourCodes;
}

export interface IColourPalette {
  primary: IColourVariations;
  secondary: IColourVariations;
  complimentary: IColourVariations;
  action: IColourVariations;
  error: IColourVariations;
  background: IColourCodes;
  surface: IColourCodes;
  white: IColourCodes;
  black: IBlackVariations;
}
