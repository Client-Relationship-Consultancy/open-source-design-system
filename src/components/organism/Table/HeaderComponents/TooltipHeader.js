import React, { Component } from "react"
import PropTypes from "prop-types"
import styled from "styled-components"

import Icon from "atoms/Icon"
import Tooltip from "atoms/Tooltip"

const Container = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
`

const nextSortOrder = {
  asc: "desc",
  desc: "noSort",
  noSort: "asc",
}

const sortIcon = {
  asc: "long-arrow-alt-up",
  desc: "long-arrow-alt-down",
}

export const getCurrentSortIcon = (sort, theme) =>
  sortIcon[sort] && <Icon name={sortIcon[sort]} color="black" shade="tint60" theme={theme} />

class TooltipHeader extends Component {
  state = {
    sort: "noSort",
    hovered: false,
  }

  handleMouseEnter = () => this.setState({ hovered: true })

  handleMouseLeave = () => this.setState({ hovered: false })

  onMenuClicked = () => this.props.showColumnMenu(this.menuButton)

  onSortChanged = () => {
    this.setState(
      prevState => ({ sort: nextSortOrder[prevState.sort] }),
      () => {
        const order = this.state.sort === "noSort" ? "" : this.state.sort
        this.props.setSort(order)
      },
    )
  }

  getHeaderNameWithTooltip = () =>
    this.state.hovered ? (
      <Tooltip tooltipContent={this.props.tooltipContent} id={`tooltip-${this.props.column.colId}`}>
        {this.props.displayName}
      </Tooltip>
    ) : (
      this.props.displayName
    )

  render() {
    const { column } = this.props
    const { colDef } = column
    const { hovered } = this.state
    return (
      <Container
        onClick={colDef.sortable && this.onSortChanged}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
        id={this.props.id}
        className={this.props.className}
      >
        <div>{this.getHeaderNameWithTooltip()}</div>
        {colDef.sortable && getCurrentSortIcon(this.state.sort, this.props.agGridReact.props.theme)}
        {colDef.filter && hovered && (
          <div
            ref={menuButton => {
              this.menuButton = menuButton
            }}
          >
            <Icon
              onClick={this.onMenuClicked}
              theme={this.props.agGridReact.props.theme}
              name="bars"
              color="black"
              shade="tint60"
            />
          </div>
        )}
      </Container>
    )
  }
}

TooltipHeader.displayName = "TooltipHeader"

TooltipHeader.propTypes = {
  displayName: PropTypes.string.isRequired,
  tooltipContent: PropTypes.string.isRequired,
  column: PropTypes.object,
  showColumnMenu: PropTypes.func,
  setSort: PropTypes.func,
  id: PropTypes.string,
  className: PropTypes.string,
  agGridReact: PropTypes.shape({
    props: PropTypes.shape({
      theme: PropTypes.object,
    }),
  }),
}
export default TooltipHeader
