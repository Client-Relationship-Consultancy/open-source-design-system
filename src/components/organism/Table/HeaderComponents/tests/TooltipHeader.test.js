import React from "react"
import { shallow } from "enzyme"
import { colourPalette } from "../../../../../brandColours"
import "jest-styled-components"
import TooltipHeader, { getCurrentSortIcon } from "../TooltipHeader"

const theme = {
  props: {
    theme: colourPalette,
  },
}

describe("TooltipHeader test", () => {
  const mockShowColumnMenu = jest.fn()
  const mockSetSort = jest.fn()
  const component = shallow(
    <TooltipHeader
      displayName="First Name"
      tooltipContent="Tooltip hey"
      column={{ colDef: { sortable: true, filterable: true } }}
      showColumnMenu={mockShowColumnMenu}
      setSort={mockSetSort}
      agGridReact={theme}
    />,
  )

  it("match last snapshot", () => {
    expect(component).toMatchSnapshot()
  })

  it("getCurrentSortIcon - asc", () => {
    const asc = shallow(getCurrentSortIcon("asc"))
    expect(asc).toMatchSnapshot()
  })

  it("getCurrentSortIcon - desc", () => {
    const desc = shallow(getCurrentSortIcon("desc"))
    expect(desc).toMatchSnapshot()
  })

  it("should handle onMenuClicked", () => {
    component.instance().onMenuClicked()
    expect(mockShowColumnMenu).toHaveBeenCalled()
  })

  it("should handle onSortChanged in the correct order", () => {
    // Initial state
    expect(component.state().sort).toBe("noSort")

    // First sort
    component.instance().onSortChanged()
    expect(mockSetSort).toHaveBeenCalledWith("asc")
    expect(component.state().sort).toBe("asc")

    // Second sort
    component.instance().onSortChanged()
    expect(mockSetSort).toHaveBeenCalledWith("desc")
    expect(component.state().sort).toBe("desc")

    // Third sort
    component.instance().onSortChanged()
    expect(mockSetSort).toHaveBeenCalledWith("")
    expect(component.state().sort).toBe("noSort")
  })

  it("should handle onMouseEnter and show the tooltip", () => {
    component.simulate("mouseenter")
    expect(component.instance().state.hovered).toBe(true)
    expect(component).toMatchSnapshot()
  })

  it("should handle onMouseLeave and remove the tooltip", () => {
    component.simulate("mouseleave")
    expect(component.instance().state.hovered).toBe(false)
  })
})
