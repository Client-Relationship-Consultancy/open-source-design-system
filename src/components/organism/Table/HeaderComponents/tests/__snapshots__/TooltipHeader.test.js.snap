// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TooltipHeader test getCurrentSortIcon - asc 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`TooltipHeader test getCurrentSortIcon - desc 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`TooltipHeader test match last snapshot 1`] = `
<TooltipHeader__Container
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div>
    First Name
  </div>
</TooltipHeader__Container>
`;

exports[`TooltipHeader test should handle onMouseEnter and show the tooltip 1`] = `
<TooltipHeader__Container
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div>
    <Tooltip
      id="tooltip-undefined"
      showIcon={true}
      tooltipContent="Tooltip hey"
    >
      First Name
    </Tooltip>
  </div>
</TooltipHeader__Container>
`;
